language/move-vm/types/src/native_functions/context.rs

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    interpreter_context::InterpreterContext,
+    loaded_data::{runtime_types::Type, types::FatType},
+    values::Struct,
+};
+use libra_types::{account_address::AccountAddress, language_storage::ModuleId};
+use move_core_types::{gas_schedule::CostTable, identifier::IdentStr};
+use std::fmt::Write;
+use vm::errors::VMResult;
+
+/// Native function context.
+pub trait NativeContext {
```

Comment:
`NativeContext` feels like a very rich API and it's not clear what the reasons are.
You could imagine `NativeContext` to expose only 2 APIs: `CostTable` and `DataStore`.
- `CostTable` is the API every native would need and the **only** API all native functions today need (and those we have in mind)
- `DataStore` is the API that would have to be implemented in order to interact with a data store. `save_data()` and `save_event()` would be the 2 API needed in the Libra case. `DataStore` is tricky because it has to work in concert with the API used for all bytecodes operating on global data.

Moreover `DataStore` seems a bit too special/powerful and we could define a different or parallel way to expose it.
In other words we could imagine not giving `DataStore` to every single native function.

Libra could be made to work with those 2 "concepts" only but we are wondering what your use cases and scenarios are. We understand native functions are a way to expose very tricky behavior and it is not too easy to understand where to draw the line.

Thoughts?

Reply:
Yeah. Maybe it's the wrong idea to row all under one API.
I think we should impoverish the input function. (ty_args, arguments) to reduce the number of input parameters. 
And NativeContext should be divided into data sources and context with CostTable and other useful things.
One more important thing is that we would like to be able to deduct gas before executing the body function with the macro gas! 
Two points:
1) the function can be heavy and it might not make sense to spend resources on its execution if there is still not enough gas.
2) The current interface (VMResult<NativeResult>) does not allow to simply use '?' because in this case gas is not deduct.